utils/to_string.hh

```diff
@@ -107,45 +90,23 @@ std::ostream& operator<<(std::ostream& os, const std::tuple<T...>& p) {
     return print_tuple(os, p, std::make_index_sequence<sizeof...(T)>());
 }
 
-template <typename T, typename... Args>
-std::ostream& operator<<(std::ostream& os, const std::unordered_set<T, Args...>& items) {
-    os << "{" << utils::join(", ", items) << "}";
-    return os;
-}
-
-template <typename T>
-std::ostream& operator<<(std::ostream& os, const std::set<T>& items) {
-    os << "{" << utils::join(", ", items) << "}";
-    return os;
-}
-
-template<typename T, size_t N>
-std::ostream& operator<<(std::ostream& os, const std::array<T, N>& items) {
-    os << "{" << utils::join(", ", items) << "}";
-    return os;
-}
-
-template <typename K, typename V, typename... Args>
-std::ostream& operator<<(std::ostream& os, const std::unordered_map<K, V, Args...>& items) {
-    os << "{" << utils::join(", ", items) << "}";
+// Exclude string-like types to avoid printing them as vector of chars
+template <std::ranges::range Range>
+requires (
+       !std::convertible_to<Range, std::string>
+    && !std::convertible_to<Range, std::string_view>
+    && !std::convertible_to<Range, bytes>
+    && !std::convertible_to<Range, bytes_view>
+    && !std::same_as<Range, boost::unit_test::basic_cstring<const char>>
```

Comment:
Again, I don't like the boost-unit-test specific code here. Why do we need it? Can't we figure out a different way to do this without relying on the specific boost type (e.g., maybe another convertible_to trick?)

Reply:
`boost::unit_test::basic_cstring` (a.k.a `cstring`) is used in boost::test for the test names and other places. it has its own `operator<<`. to avoid printing it as a range, and to avoid referencing it in `to_string.h`, probably we can have a trait class specialized for it and other range-alike types, as an alternative to type constraints? the trait class for `boost::unit_test::basic_cstring` can be hosted by another header file under `test/` ? but i am wondering what is the use case that we want to format the `cstring` with `seastar::format()`, maybe another option is that we could expose the `cstring` using `seastar::testing::seastar_test` via a `sstring` or a `std::string` ?

Reply:
We don't format boost_test cstrings using seastar::format.
I think that the compiler picks up our operator<< since they conform to the `std::ranges::range` concept and it calls it implicitly.  Hence this weird exclusion.